projects/ui-components/lf-metadata/lf-field-adhoc-container/lf-field-add-remove/lf-field-add-remove.component.html

```diff
@@ -8,33 +8,39 @@
       <div class="ms-font-m back-button-text">{{ADD_REMOVE_FIELDS | async}}</div>
   </div>
   <div class="back-button-padding"></div>
-  <div class="field-picker-container" [ngClass]="{'has-changes': areCheckboxChanges}">
-      <div class="search">
-          <input [value]="fieldFilterText" [formControl]="filterFieldsControl" placeholder="{{SEARCH_FIELDS | async}}"
-              class="searchBox">
-          <div class="icon icon-search"></div>
-          <div *ngIf="fieldFilterText && fieldFilterText.length > 0" (click)="onClearFields()"
-              class="icon icon-clear">
+  <div class="add-remove-content">
+    <lf-loader-component *ngIf="isLoading" class="lf-adhoc-loader"></lf-loader-component>
+    <div *ngIf="shouldShowErrorMessage" class="lf-adhoc-loader error-entries-text">{{ AN_ERROR_OCCURED | async }}</div>
+    <div class="field-picker-container" *ngIf="!isLoading && !shouldShowErrorMessage" [ngClass]="{'has-changes': areCheckboxChanges}">
+        <div class="search">
+            <input [value]="fieldFilterText" [formControl]="filterFieldsControl" placeholder="{{SEARCH_FIELDS | async}}"
+                class="searchBox">
+            <div class="icon icon-search"></div>
+            <div *ngIf="fieldFilterText && fieldFilterText.length > 0" (click)="onClearFields()"
+                class="icon icon-clear">
+            </div>
+        </div>
+        <cdk-virtual-scroll-viewport itemSize="25" class="pick-field">
+          <div *cdkVirtualFor="let field of displayFieldInfos, let i = index">
+              <mat-checkbox color="primary" class="lf-checkbox" (change)="onUpdateCheckbox($event, field)" id=`add-remove-checkbox-{{i}}`
+                  [checked]="isFieldSelected(field)" title="{{field.name}}" [disabled]="field.inTemplateSelected">
+                  {{field.name}} ({{field | getFieldType: LOCALIZED_FIELD_TYPES | async}})
+              </mat-checkbox>
           </div>
-      </div>
-      <div *ngFor="let field of displayFieldInfos, let i = index" class="pick-field">
-          <mat-checkbox color="primary" class="lf-checkbox" (change)="onUpdateCheckbox($event, field)" id=`add-remove-checkbox-{{i}}`
-              [checked]="isFieldSelected(field)" title="{{field.name}}" [disabled]="field.inTemplateSelected">
-              {{field.name}} ({{field | getFieldType: LOCALIZED_FIELD_TYPES | async}})
-          </mat-checkbox>
-      </div>
-      <div class="fields-empty" *ngIf="!displayFieldInfos || displayFieldInfos.length===0">
-          {{NO_MATCHING_FIELDS_FOUND | async}}
-      </div>
+      </cdk-virtual-scroll-viewport>
+        <div class="fields-empty" *ngIf="!displayFieldInfos || displayFieldInfos.length===0">
```

Comment:
there is a bug where if the repo has no fields this should say no available fields instead.. Is there a way to tell that? allFieldInfos vs displayed?
We have a separate bug for that, just something to keep in mind